python/taichi/lang/stmt_builder.py

```diff
@@ -0,0 +1,675 @@
+                        "Function definition not allowed in 'ti.func'.")
+            if impl.get_runtime().experimental_real_function:
+                transform_as_kernel()
+            else:
+                # Transform as func (all parameters passed by value)
+                arg_decls = []
+                for i, arg in enumerate(args.args):
+                    # Directly pass in template arguments,
+                    # such as class instances ("self"), fields, SNodes, etc.
+                    if isinstance(ctx.func.argument_annotations[i],
+                                  ti.template):
+                        continue
+                    # Create a copy for non-template arguments,
+                    # so that they are passed by value.
+                    arg_init = parse_stmt('x = ti.expr_init_func(0)')
+                    arg_init.targets[0].id = arg.arg
+                    ctx.create_variable(arg.arg)
+                    arg_init.value.args[0] = parse_expr(arg.arg +
+                                                        '_by_value__')
+                    args.args[i].arg += '_by_value__'
+                    arg_decls.append(arg_init)
+
+        with ctx.variable_scope():
+            node.body = build_stmts(ctx, node.body)
+
+        node.body = arg_decls + node.body
+        node.body = [parse_stmt('import taichi as ti')] + node.body
+        return node
+
+    @staticmethod
+    def build_Return(ctx, node):
+        node.value = build_expr(ctx, node.value)
+        if ctx.is_kernel or impl.get_runtime().experimental_real_function:
+            # TODO: check if it's at the end of a kernel, throw TaichiSyntaxError if not
+            if node.value is not None:
+                if ctx.returns is None:
+                    raise TaichiSyntaxError(
+                        f'A {"kernel" if ctx.is_kernel else "function"} '
+                        'with a return value must be annotated '
+                        'with a return type, e.g. def func() -> ti.f32')
+                ret_expr = parse_expr('ti.cast(ti.Expr(0), 0)')
+                ret_expr.args[0].args[0] = node.value
+                ret_expr.args[1] = ctx.returns
+                ret_stmt = parse_stmt('ti.core.create_kernel_return(ret.ptr)')
+                # For args[0], it is an ast.Attribute, because it loads the
+                # attribute, |ptr|, of the expression |ret_expr|. Therefore we
+                # only need to replace the object part, i.e. args[0].value
+                ret_stmt.value.args[0].value = ret_expr
+                return ret_stmt
+        return node
+
+    @staticmethod
+    def build_Module(ctx, node):
+        with ctx.variable_scope():
+            # Do NOT use |build_stmts| which inserts 'del' statements to the
+            # end and deletes parameters passed into the module
+            node.body = [build_stmt(ctx, stmt) for stmt in list(node.body)]
+        return node
+
+    @staticmethod
+    def build_Global(ctx, node):
```

Comment:
Note that the behavior of `global` is different from transformer.py -- If we use `global`, the program may crash later before this PR.